README.md
# rk.FactorAnalysis

This R package provides a GUI to conduct principal component and factor analysis.

![Parallel analysis (Horn)](./inst/screenshots/parallel_analysis.png?raw=true "rk.FactorAnalysis: Parallel analysis")

It is a plugin package for [RKWard](https://rkward.kde.org), a powerful GUI and IDE for [R](https://r-project.org),
and will **only work in conjunction with RKWard**. However, RKWard is free software, please
[check it out](https://rkward.kde.org).

## Installation

### Installation via RKWard

After you installed RKWard, simply install this package like any other R package. The easiest way to get the most
recent stable release is to use RKWard's package management:

- Go to `Settings` -> `Manage R packages and plugins...` -> `Install/Update/Remove R packages`
- Check `Show only packages providing RKWard dialogs` and look at `New Packages`
- Check the desired package(s) and hit `Apply`

After you installed the package, RKWard should immediately have additional menu entries -- there's
usually no need to restart the application.

### Installation via GitHub

To install it directly from GitHub, you can use `install_github()` from the [devtools](https://github.com/hadley/devtools) package:

```
library(devtools)
install_github("rkward-community/rk.FactorAnalysis") # stable release
install_github("rkward-community/rk.FactorAnalysis", ref="develop") # development release
```

You can safely ignore warnings about the (unavailable) package "rkward", it was installed together with RKWard.
 
## Contributing

See https://docs.kde.org/trunk5/en/rkward/rkwardplugins/index.html for documentation on writing plugins for RKWard.
Please contact the [RKWard development mailing list](https://mail.kde.org/mailman/listinfo/rkward-devel)
for help, reports and requests.

- Submit pull requests or patches
- Contact us to get write access

This plugin is built using the [rkwarddev](https://files.kde.org/rkward/R/pckg/rkwarddev/index.html) package. This means that its development is mainly focussed on one
script file, located at `rk.FactorAnalysis/inst/rkward/rkwarddev_plugin_script_FA.R`. This script is written in `R` code and *generates*
all of the other plugin files when run. Therefore, in case you'd like to add to this plugin, please add to this script file.

### Branches

Please note that all development happens in the `develop` branch. Pull requests against the `master` branch will be rejected, as it is reserved for the current stable release.

## Licence

Copyright 2012-2022 Meik Michalke <meik.michalke@hhu.de>

rk.FactorAnalysis is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

rk.FactorAnalysis is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with rk.FactorAnalysis.  If not, see <http://www.gnu.org/licenses/>.


ChangeLog
ChangeLog for package rk.FactorAnalysis

changes in version 0.2-1 (2022-05-03)
unreleased:
  - this release is under development

changes in version 0.01-16 (2022-05-03)
fixed:
  - screenshot URL and removed flattr-link in README.md

changes in version 0.01-15 (2017-02-25)
fixed:
  - fixed a missing comma in the printout code
  - fixed factor loadings highlighting (now also respects negative loadings)
added:
  - new dialog to calculate measure of sampling adequacy (MSA,
    Kaiser-Meyer-Olkin)
changed:
  - updated rk.header() code to newer style
  - the single script file was split into one file for each component
  - adjusted the generator script to implement the new preview approach,
    using "printout" and dropping "doPrintout"
  - updated german tralsation

changes in version 0.01-14 (2015-12-18)
added:
  - plugin is now translatable (see inst/rkward/po/)
  - german translation
changed:
  - updated the generator script
  - rebuilt with rkwarddev 0.08-1

changes in version 0.01-13 (2015-03-14)
fixed:
  - replaced the outdated links to sourceforge with our new KDE URL

changes in version 0.01-12 (2015-03-13)
changed:
  - rebuilt using rkwarddev 0.07-1

changes in version 0.01-11 (2013-04-04)
added:
  - debianized the plugin
fixed:
  - removed erroneous pluginmap inclusions from XML files

changes in version 0.01-10 (2013-02-24)
changed:
  - updated rkwarddev generator script (compatibility with 0.06-2) and
    regenerated plugin code

changes in version 0.01-9 (2012-10-08)
fixed:
  - bug in testing for rotation method led to errors when trying to compute
    variance explained when orthogonal rotation was used with PCA. the result
    was that no real output was generated, only an error message
added:
  - added VSS/MAP dialog
changed:
  - moved parallel analysis and scree plot from "plots" to "analysis"

changes in version 0.01-8 (2012-09-29)
fixed:
  - added generic pre/post plot options
  - ChangeLog now shows the correct package name...
  - a typo (thx birk ;-))

changes in version 0.01-7 (2012-08-26)
fixed:
  - scree plot dialog referenced the wrong varselector
added:
  - dialog for Horn's parallel analysis plots

changes in version 0.01-6 (2012-04-07)
changes:
  - added a ChangeLog file
  - rebuild package for R 2.15



inst/rkward/rkwarddev_plugin_script_FA.R
# the plugin code was generated by this script
# you should not change the plugin code directly, but this script
# note: this script only creates objects in your workspace,
# *EXCEPT* for the last call, see below.

require(rkwarddev)
rkwarddev.required("0.08-2")

local({
  # set the output directory to overwrite the actual plugin
  output.dir <- tempdir()
  overwrite <- TRUE
  # if you set guess.getters to TRUE, the resulting code will need RKWard >= 0.6.0
  guess.getter <- TRUE
  rk.set.indent(by="  ")
  rk.set.empty.e(TRUE)
  update.translations <- TRUE

  ## you must first setwd() to $YOURSOURCES/rk.FactorAnalysis/inst/rkward for this to run properly
  rkfa.components.root <- normalizePath("rkwarddev_components")

  about.info <- rk.XML.about(
    name="rk.FactorAnalysis",
    author=c(
      person(
        given="Meik",
        family="Michalke",
        email="meik.michalke@hhu.de",
        role=c("aut","cre")
      )
    ),
    about=list(
      desc="RKWard GUI to conduct principal component and factor analysis",
      version="0.2-1",
      url="https://rkward.kde.org",
      long.desc="RKWard GUI to conduct principal component and factor analysis (using the psych package). Also includes dialogs for scree plots, correlation plots, VSS/MAP and parallel analysis."
    )
  )

  dependencies.info <- rk.XML.dependencies(
    dependencies=list(
      rkward.min="0.6.5"
    ),
    package=list(
      c(name="psych", min="1.1.10")
    )
  )


  ## Components
  source(file.path(rkfa.components.root, "component_PCA_FA.R"), local=TRUE)
  source(file.path(rkfa.components.root, "component_scree_plot.R"), local=TRUE)
  source(file.path(rkfa.components.root, "component_parallel_analysis.R"), local=TRUE)
  source(file.path(rkfa.components.root, "component_VSS_MAP.R"), local=TRUE)
  source(file.path(rkfa.components.root, "component_correlation_plot.R"), local=TRUE)
  source(file.path(rkfa.components.root, "component_KMO.R"), local=TRUE)
  # source(file.path(rkfa.components.root, "component_.R"), local=TRUE)


  #############
  ## if you run the following function call, files will be written to tempdir!
  #############
  # this is where it get's serious, that is, here all of the above is put together into one plugin

  rk.FactorAnalysis.dir <<- rk.plugin.skeleton(
    about.info,
    path=output.dir,
    guess.getter=guess.getter,
    xml=list(
      logic=lgc.sect,
      dialog=full.dialog
    ),
    js=list(
      globals="var isObrot;",
      results.header=FALSE,
      require="psych",
      calculate=js.calc,
      printout=js.print
    ),
    pluginmap=list(name="Factor analysis", hierarchy=list("analysis", "Factor analysis")),
    components=list(
      scree.component,
      crplt.component,
      prll.component,
      vss.component,
      KMO.component
    ),
    dependencies=dependencies.info,
    create=c("pmap", "xml", "js", "desc"),
    overwrite=overwrite,
    tests=FALSE,
  #  edit=TRUE,
  #  show=TRUE,
    load=TRUE,
    hints=FALSE
  )

  if(isTRUE(update.translations)){
    rk.updatePluginMessages(file.path(output.dir,"rk.FactorAnalysis","inst","rkward","rk.FactorAnalysis.pluginmap"))
  } else {}
})
